### PR TITLE
Fix #723 Calculated Columns - Replace PR 734

### LINKED ISSUE #723: Feature request for Excel tables: Calculated Column Formula
Microsoft Excel has a powerful feature for [tables](https://support.microsoft.com/en-us/office/overview-of-excel-tables-7ab0bb7d-3a9e-4b56-a3c9-6c94334e492c), named the [Calculated Column Formulas](https://support.microsoft.com/en-us/office/use-calculated-columns-in-an-excel-table-873fbac6-7110-4300-8f6f-aafa2ea11ce8):

1. Entering or changing the formula in one cell automatically applies it to all the cells in the same column
2. Resizing the table or inserting a row automatically replicates the calculated column formulas to these new rows
3. The formulas may use the names of the columns instead of classic range references
4. It's still possible to have other values in the column but is marked as an error ("This is generally not recommended though").

Currently, abap2xlsx:

1. may create a table via the method BIND_TABLE (only), the values and formulas of the table are passed via an internal table, another parameter is the field catalog which indicates which columns contain values and which others contain formulas.
2. a repeated formula must be entered in every line of the internal table. Hence the same formula `TblFlights[[#This Row],[Airfare]]+100` can be defined in every row (formula using "[structured references](https://support.microsoft.com/en-us/office/using-structured-references-with-excel-tables-f5ed2452-2337-4f71-bed3-c8ae6d2b276e)"), while a formula using cell references `D2+100` must vary at every row (i.e. the formula in the next row must be `D2+100`).
3. has no demo program of formulas with BIND_TABLE.
4. does not have a way to produce an Excel file where the user can have the calculated formulas automatically replicated when he/she resizes the table
5. does not have a way to hide cell errors (which is especially useful for a calculated column if one cell needs to have a different value).

I propose these features:

1. enter a formula only once via the field catalog (instead of providing/repeating the formula in all lines of the internal table), whatever the formula is of the form `TblFlights[[#This Row],[Airfare]]+100` or `D2+100`
2. provide a demo program of formulas with BIND_TABLE
3. produce an Excel file where the user can have the calculated formulas automatically replicated when he/she resizes the table
4. produce an Excel file where the cell errors can be hidden

NB: it's not possible to enter `[@Airfare]`, only the long formula is permitted `TblFlights[[#This Row],[Airfare]]`.

Illustrations from Microsoft Web site (links above):
image

### PR DESCRIPTION
**Pull request by sandraros**
Fix #723. Currently I can do only the WRITER part, because the READER currently doesn't read the Tables at all, just its contents, i.e. the Table generated by BIND_TABLE doesn't have a counterpart in the READER.

### CODE REVIEW

**Comment by AndreaBorgia-Abo** [comment]
Same as #839 : need Excel proper to test results.

**Comment by AndreaBorgia-Abo** [comment]
The file generated by ZDEMO_EXCEL47 contains an error and was repaired by Excel 2013 (Windows) upon opening, this is how it looked like afterwards:
immagine

Here is the correction log:
```
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<recoveryLog xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main">
	<logFileName>error096120_01.xml</logFileName>
	<summary>Rilevati errori nel file 'C:\temp\47_ColumnFormulas.xlsx'</summary>
	<removedRecords summary="Di seguito è riportato un elenco dei record rimossi:">
		<removedRecord>Record rimossi: Formula dalla parte /xl/worksheets/sheet1.xml</removedRecord>
		<removedRecord>Record rimossi: Tabella dalla parte /xl/tables/table1.xml (Tabella)</removedRecord>
	</removedRecords>
</recoveryLog>
```

**Comment by sandraros** [comment]
Thanks. The problem is due to the Excel version, the feature works only from Excel 365 or Excel 2021 (https://support.microsoft.com/en-us/office/use-calculated-columns-in-an-excel-table-873fbac6-7110-4300-8f6f-aafa2ea11ce8).
image
image
image

Is it okay to add the required version in the title or should we do more? (maybe like defining a target Excel version while instantiating `ZCL_EXCEL` so that to raise an exception with the right message if the program attempts using some unsupported feature)
image

File: src/demos/zdemo_excel47.prog.xml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    <DBNA>D$</DBNA>
+    <SUBC>1</SUBC>
+    <FIXPT>X</FIXPT>
+    <LDBNAME>D$S</LDBNAME>
+    <UCCHECK>X</UCCHECK>
+   </PROGDIR>
+   <TPOOL>
+    <item>
+     <ID>R</ID>
+     <ENTRY>abap2xlsx Demo: Table Calculated Columns</ENTRY>
```
Comment (AndreaBorgia-Abo) [inline src/demos/zdemo_excel47.prog.xml]: 2016 and above

**Comment by sandraros** [comment]
Hmm sorry. Yes. And also valid for Excel 2013:
> Excel for Microsoft 365 Excel for Microsoft 365 for Mac Excel 2021 Excel 2021 for Mac Excel 2019 Excel 2019 for Mac Excel 2016 Excel 2016 for Mac Excel 2013 Excel 2010 Excel 2007 Excel for Mac 2011

The problem about column "C4." is due to `IFS` (https://support.microsoft.com/en-us/office/ifs-function-36329a26-37b2-467c-972b-4a39bd951d45):
> Excel for Microsoft 365 Excel for Microsoft 365 for Mac Excel for the web Excel 2021 Excel 2021 for Mac Excel 2019 Excel 2019 for Mac

The problem with columns "C6." and "C7." is due to `FILTER` (https://support.microsoft.com/en-us/office/filter-function-f4f7cb66-82eb-4767-8f7c-4877ad80c759):
> Excel for Microsoft 365 Excel for Microsoft 365 for Mac Excel for the web Excel 2021 Excel 2021 for Mac Excel for iPad Excel for iPhone Excel for Android tablets Excel for Android phones

What do you think?

**Comment by AndreaBorgia-Abo** [comment]
Right, I asked a colleague and indeed Excel 2016 has no issues. The way I read the "calculated columns" list of versions is that it's pretty Mac-specific and I am using Windows.
I believe we could do both: edit the title (and maybe a note in the code with links) in this demo (to close this PR) and then open a bug to track the Excel version issue: I expect this kind of glitches could happen a lot but for a proper fix we'd need keep tables of applicability lists for each feature and have a master version switch against which to compare.
Unrelated question: why some PR have a branch on the main repository and some on yours? Just curious :)
Let me finish the review, we'll see...

**Comment by sandraros** [comment]
Agreed for "applicability list". Let's close the PR first. Someone told me that Excel 2016 doesn't exist, it's the internal number, probably the official name is Excel 365 or Excel 2019 (https://en.wikipedia.org/wiki/Microsoft_Excel#Excel_2019,_Office_365_and_subsequent_(v16.0)). Each supported version is both for Mac and for Windows. I propose the text "(from Excel 365 or 2019)".
I now create a branch directly in abap2xlsx, the advantage is that it's automatically deleted when you merge the PR (it's what Lars does). I still have old PRs/branches in my own repository, I hope one day I can clean it.

**Review by AndreaBorgia-Abo — APPROVED** [review]

**Comment by AndreaBorgia-Abo** [comment]
Hmm, I prefer to keep branches on my own repo to minimize the risk of pushing something to master outisde the PR channel.
Crucially, I also get a prompt after merging to delete the branch here on GH. At the next pull it will be gone from the demo system as well.
Anyway, merging :)

**Pull request merged**